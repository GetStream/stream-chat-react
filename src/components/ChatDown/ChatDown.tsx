import React from 'react';

import { LoadingChannels } from '../Loading/LoadingChannels';

import placeholder from '../../assets/str-chat__connection-error.svg';
import { useTranslationContext } from '../../context/TranslationContext';

export type ChatDownProps = {
  /** The type of error */
  type: string;
  /** The image url for this error */
  image?: string;
  /** The error message to show */
  text?: string;
};

const UnMemoizedChatDown: React.FC<ChatDownProps> = (props) => {
  const { image, text, type = 'Error' } = props;

  const { t } = useTranslationContext('ChatDown');

  return (
    <div className='str-chat__down'>
      <LoadingChannels />
      <div className='str-chat__down-main'>
        <img alt='Connection error' data-testid='chatdown-img' src={image || placeholder} />
        <h1>{type}</h1>
        <h3 aria-live='assertive'>
          {text || t('Error connecting to chat, refresh the page to try again.')}
        </h3>
      </div>
    </div>
  );
};

/**
 * A simple indicator that chat functionality isn't available, triggered when the Chat API is unavailable or your network isn't working.
 */
export const ChatDown = React.memo(UnMemoizedChatDown) as typeof UnMemoizedChatDown;
