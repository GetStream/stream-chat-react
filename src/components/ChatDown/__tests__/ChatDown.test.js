import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ChatDown } from '../ChatDown';

describe('ChatDown', () => {
  afterEach(cleanup);
  it('should render component with its default props', () => {
    const tree = renderer.create(<ChatDown />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render component with custom text', async () => {
    const text = 'custom text';
    const { getByText } = render(<ChatDown text={text} />);

    await waitFor(() => {
      expect(getByText(text)).toBeInTheDocument();
    });
  });

  it('should render component with custom image url', async () => {
    const image = 'https://random.url/image.png';
    const Component = <ChatDown image={image} />;
    const { getByTestId } = render(Component);

    await waitFor(() => {
      expect(getByTestId('chatdown-img')).toHaveAttribute('src', image);
    });

    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render custom image component', async () => {
    const ImageComponent = () => <div data-testid={'chatdown-img-custom'}>custom</div>;
    const Component = <ChatDown image={<ImageComponent />} />;
    const { getByTestId } = render(Component);

    await waitFor(() => {
      expect(getByTestId('chatdown-img-custom')).toBeInTheDocument();
    });

    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render component with custom type', async () => {
    const type = 'Warning';
    const { getByText } = render(<ChatDown type={type} />);

    await waitFor(() => {
      expect(getByText(type)).toBeInTheDocument();
    });
  });
});
