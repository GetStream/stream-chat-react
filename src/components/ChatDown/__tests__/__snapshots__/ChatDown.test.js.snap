// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatDown should render component with custom image url 1`] = `
<div
  className="str-chat__down"
>
  <div
    className="str-chat__loading-channels"
  >
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
  </div>
  <div
    className="str-chat__down-main"
  >
    <img
      alt="Connection error"
      data-testid="chatdown-img"
      src="https://random.url/image.png"
    />
    <h1>
      Error
    </h1>
    <h3
      aria-live="assertive"
    >
      Error connecting to chat, refresh the page to try again.
    </h3>
  </div>
</div>
`;

exports[`ChatDown should render component with its default props 1`] = `
<div
  className="str-chat__down"
>
  <div
    className="str-chat__loading-channels"
  >
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
    <div
      className="str-chat__loading-channels-item"
    >
      <div
        className="str-chat__loading-channels-avatar"
      />
      <div
        className="str-chat__loading-channels-meta"
      >
        <div
          className="str-chat__loading-channels-username"
        />
        <div
          className="str-chat__loading-channels-status"
        />
      </div>
    </div>
  </div>
  <div
    className="str-chat__down-main"
  >
    <img
      alt="Connection error"
      data-testid="chatdown-img"
      src={
        Object {
          "process": [Function],
        }
      }
    />
    <h1>
      Error
    </h1>
    <h3
      aria-live="assertive"
    >
      Error connecting to chat, refresh the page to try again.
    </h3>
  </div>
</div>
`;
